lib/chef/provider/package.rb

```diff
@@ -452,9 +452,15 @@ def target_version_array
                 elsif current_version.nil?
                   logger.trace("#{new_resource} #{package_name} not installed, installing #{candidate_version}")
                   target_version_array.push(candidate_version)
-                else
+                elsif version_compare(current_version, candidate_version) == 0
                   logger.trace("#{new_resource} #{package_name} #{current_version} already installed")
                   target_version_array.push(nil)
+                elsif !allow_downgrade && version_compare(current_version, candidate_version) == 1
+                  logger.trace("#{new_resource} #{package_name} has installed version #{current_version}, which is newer than available version #{candidate_version}. Skipping...)")
+                  target_version_array.push(nil)
+                else
```

Comment:
`:install` doesn't upgrade like this. Unless the version is specified, `:install` simply installs if nothing is installed. You need to use `:upgrade` to upgrade-to-latest or `:install` with a version to move to a new version.

Reply:
Thanks  for looking at this.  Also, thanks for the description of the required behavior (I'm new to chef and didn't know this).

I added this code to fix the following test case:
---8<---
package 'empty-test-rpm-version-install' do
version '1:2.02-78.el8_1.1'
action :install
end
---8<---

With this change, new_version (which contained the user specified version) has been nil'd out (by the dnf subclass).  So I no longer have the context necessary to make the determination if we should do the upgrade in the scenario you've outlined.  :(

I'm hesitant to update package.rb to add special handling for this case since it's a generic resource handler and this information is only needed for this one dnf edge case.

My first idea for a fix would probably be to update the dnf version of new_version_array() such that: instead of always nil'ing out the entire array, if we're dealing with an install package resource, then we could set new_version to the available_version if both are non-nil (and set it to nil otherwise).

I'd be curious to get @lamont-granquist perspective on this idea (and the rest of the diff).  I noticed that he just closed the issue this diff is addressing as a dup of another issue+PR, so I suspect this PR may be obsoleted shortly.

Reply:
I've updated the PR with the changes I described above, which allowed me to revert all the updates to provider/package.rb (so I've restored the old install behavior).  I've verified that all my original tests (from the comments) pass.